website/blog/2023-09-26-http-streaming-back-pressure-and-cancellation/index.md

```diff
@@ -0,0 +1,259 @@
+
+![cancellation](./cancellation.png)
+
+## How to handle canellation?
+
+The core idea is straightforward: on the server side, we need to listen to the `close` event and check if the connection is still valid before pulling data from the LLM stream.
+
+```JavaScript
+app.get('/api/server', async (req, res) => {
+  let isConnectionClosed = false
+
+	// 1. Listen to the close event
+  req.on('close', () => {
+    console.log('close event')
+    isConnectionClosed = true
+  })
+
+  const response = await fetch('http://localhost:8000/api/llm');
+  const reader = response.body.getReader();
+
+  const serverStream = new ReadableStream({
+    async pull(controller) {
+      const { value, done } = await reader.read();
+      if (done) {
+        controller.close();
+      } else {
+        controller.enqueue(value);
+      }
+    }
+  });
+
+  const serverReader = serverStream.getReader();
+  for (let i = 0; i < 3; i++) {
+		// 2. Check if the connection closed before pulling the data
+    if (isConnectionClosed) {
+      // Notify LLM stream to stop producing the data
+			break;
+		}
+    const { value } = await serverReader.read();
+    const chunk = new TextDecoder("utf-8").decode(value);
+    console.log(`read ${chunk}`); 
+ 
+    await sleep(1000);
+  }
+
+  res.json({})
+})
+```
+
+
+## Implement back-pressure and cancellation for Tabby
+
+In Tabby, we need to handle both back-pressure and cancellation for code completion so that we can react to users' inputs as soon as possible, while protecting the usage of the model to keep the performance.
+
+On the client side, everytime we receive a new input from a user, we need to abort the previous query to the server and fetch a new response from LLM.
+
+```JavaScript
+let controller;
+
+const callServer = (prompt) => {
+  controller = new AbortController();
+  const signal = controller.signal;
+	// 3. calling server API to get the result with the prompt
+  const response = await fetch('/v1/completions', {
+		method: "POST",
+		headers: {
+	    'Content-Type': 'application/json'
+	  },
+	  body: JSON.stringify({ prompt })
+    signal
+  });
+}
+
+const onChange = (e) => {
+	// 2. Abort the previous request
+  if (controller) controller.abort();
+  callServer(e.target.value)
+}
+
+// 1. Debounce the input 100ms for example
+<input onChange={debounce(onChange)} />
+
+```
+
+
+On the server side, we need to listen to the `close` event and tell the LLM to stop generating results.
+
+```JavaScript
+app.post('/v1/completions', async (req, res) => {
+	const { prompt } = req.body
+
+	// 1. Listen to the connection close event
+  let isConnectionClosed = false
+  req.on('close', () => {
+    isConnectionClosed = true
+  })
+
+	// 2. Get the methods from LLM
+	const { cancel, text } = llmInference(prompt)
+
+  // 3. Create a ReadableStream
+  const serverStream = new ReadableStream({
+    async pull(controller) {
+      const { value, done } = await reader.read();
```

Comment:
This code sample appears to be confusing, there is no `reader` in the context.

Reply:
@wwayne Could you please take a look at the code example? Thanks

Reply:
Thanks, I sent a PR for the suggestions above https://github.com/TabbyML/tabby/pull/496